File: examples/cpptest/src/main/cpp/subsystems/doubleflywheel/DoubleFlyWheelSubsystem.cpp
// Copyright (c) 2026 Yet Another Software Suite
// SPDX-License-Identifier: LGPL-3.0-or-later

#include "subsystems/doubleflywheel/DoubleFlyWheelSubsystem.h"

#include <frc/system/plant/DCMotor.h>
#include <frc2/command/Commands.h>
#include <units/angle.h>
#include <units/angular_velocity.h>
#include <units/current.h>
#include <units/length.h>

using namespace yams::motorcontrollers;
using namespace yams::gearing;
using Cfg = SmartMotorControllerConfig;

// Two independent Kraken X60 motors: lower (CAN 4) and upper (CAN 6).
// Each wheel can be set to a different speed, which puts asymmetric backspin/topspin
// on the game piece for spin-control shooting.
DoubleFlyWheelSubsystem::DoubleFlyWheelSubsystem() {
  // kP=1 is low because Kraken velocity control runs at 1 kHz internally via
  // MotionMagic/VelocityTorqueCurrentFOC; even a small gain produces tight tracking.
  // kI and kD left at zero -- same reasoning as single-wheel shooter.
  m_lowerConfig.WithSubsystem(this)
      .WithClosedLoopMode()
      // COAST: lets the wheel spin down between shots rather than fighting back-EMF.
      .WithIdleMode(Cfg::MotorMode::COAST)
      // 3 * 4 = 12:1 reduction between Kraken shaft and wheel shaft.
      .WithMotorGearing(MechanismGearing{GearBox::FromReductionStages({3.0, 4.0})})
      // MOI measured for the physical flywheel disk; simulation uses this for
      // accurate velocity step-response without tuning purely in software.
      .WithMOI(units::kilogram_square_meter_t{0.00029264})
      .WithFeedback(1, 0, 0)
      // Feedforward at zero for now; kV/kS should be measured on the real robot
      // using SysId before closed-loop tuning.
      .WithFeedforward(frc::SimpleMotorFeedforward<units::turns>{
          units::volt_t{0}, units::unit_t<frc::SimpleMotorFeedforward<units::turns>::kv_unit>{0},
          units::unit_t<frc::SimpleMotorFeedforward<units::turns>::ka_unit>{0}})
      .WithMotorInverted(false)
      .WithTelemetry("LowerFlyWheel", Cfg::TelemetryVerbosity::HIGH);

  m_upperConfig.WithSubsystem(this)
      .WithClosedLoopMode()
      .WithIdleMode(Cfg::MotorMode::COAST)
      .WithMotorGearing(MechanismGearing{GearBox::FromReductionStages({3.0, 4.0})})
      .WithMOI(units::kilogram_square_meter_t{0.00029264})
      .WithFeedback(1, 0, 0)
      .WithFeedforward(frc::SimpleMotorFeedforward<units::turns>{
          units::volt_t{0}, units::unit_t<frc::SimpleMotorFeedforward<units::turns>::kv_unit>{0},
          units::unit_t<frc::SimpleMotorFeedforward<units::turns>::ka_unit>{0}})
      .WithMotorInverted(false)
      .WithTelemetry("UpperFlyWheel", Cfg::TelemetryVerbosity::HIGH);

  // TalonFXWrapper takes TalonFX* (pointer), not a reference -- pass address of
  // the member objects declared in the header.
  m_lowerFlyWheel.emplace(&m_talonLower, frc::DCMotor::KrakenX60(1), &m_lowerConfig);
  m_upperFlyWheel.emplace(&m_talonUpper, frc::DCMotor::KrakenX60(1), &m_upperConfig);
}

frc2::CommandPtr DoubleFlyWheelSubsystem::SetDutyCycle(double lower, double upper) {
  return Run([this, lower, upper] {
           m_lowerFlyWheel->SetDutyCycle(lower);
           m_upperFlyWheel->SetDutyCycle(upper);
         })
      .WithName("Set Duty Cycle (Double FlyWheel)");
}

frc2::CommandPtr DoubleFlyWheelSubsystem::SetVoltage(units::volt_t lower, units::volt_t upper) {
  return Run([this, lower, upper] {
           m_lowerFlyWheel->SetVoltage(lower);
           m_upperFlyWheel->SetVoltage(upper);
         })
      .WithName("Set Voltage (Double FlyWheel)");
}

frc2::CommandPtr DoubleFlyWheelSubsystem::SetVelocity(units::degrees_per_second_t lower,
                                                      units::degrees_per_second_t upper) {
  return Run([this, lower, upper] {
    m_lowerFlyWheel->SetVelocity(lower);
    m_upperFlyWheel->SetVelocity(upper);
  });
}

frc2::CommandPtr DoubleFlyWheelSubsystem::SetVelocity(
    std::function<units::degrees_per_second_t()> lower,
    std::function<units::degrees_per_second_t()> upper) {
  return Run([this, lower, upper] {
    m_lowerFlyWheel->SetVelocity(lower());
    m_upperFlyWheel->SetVelocity(upper());
  });
}

void DoubleFlyWheelSubsystem::Periodic() {}

void DoubleFlyWheelSubsystem::SimulationPeriodic() {
  m_lowerFlyWheel->SimIterate();
  m_upperFlyWheel->SimIterate();
}

File: examples/simple_robot/src/main/java/frc/robot/subsystems/doubleflywheel/DoubleFlyWheelSubsystem.java
// Copyright (c) 2025-2026 Yet Another Software Suite
// SPDX-License-Identifier: LGPL-3.0-or-later

package frc.robot.subsystems.doubleflywheel;


import static org.wpilib.units.Units.Inches;
import static org.wpilib.units.Units.Meters;
import static org.wpilib.units.Units.Pounds;
import static org.wpilib.units.Units.RPM;

import com.ctre.phoenix6.hardware.TalonFX;
import org.wpilib.math.Pair;
import org.wpilib.math.controller.SimpleMotorFeedforward;
import org.wpilib.math.system.plant.DCMotor;
import org.wpilib.units.measure.AngularVelocity;
import org.wpilib.units.measure.Distance;
import org.wpilib.units.measure.LinearVelocity;
import org.wpilib.units.measure.Voltage;
import org.wpilib.command2.Command;
import org.wpilib.command2.SubsystemBase;
import java.util.function.Supplier;
import yams.gearing.GearBox;
import yams.gearing.MechanismGearing;
import yams.motorcontrollers.SmartMotorController;
import yams.motorcontrollers.SmartMotorControllerConfig;
import yams.motorcontrollers.SmartMotorControllerConfig.ControlMode;
import yams.motorcontrollers.SmartMotorControllerConfig.MotorMode;
import yams.motorcontrollers.remote.TalonFXWrapper;

/**
 * With 2 {@link SmartMotorController}s we can control a DoubleFlyWheelSubsystem which can accurately control arc of an
 * object given tuning data.
 *
 */
public class DoubleFlyWheelSubsystem extends SubsystemBase
{
  private SmartMotorControllerConfig lowerFlyWheelConfig = new SmartMotorControllerConfig(this)
      .withControlMode(ControlMode.CLOSED_LOOP)
      .withIdleMode(MotorMode.COAST)
//      .withWheelDiameter(Inches.of(4)) // Only needed to find the MPH of the flywheel for fun.
      .withGearing(new MechanismGearing(GearBox.fromReductionStages(3, 4)))
      .withMomentOfInertia(Inches.of(4), Pounds.of(2))
      .withClosedLoopController(1,
                                0,
                                0) // You generally do not want a profile because its not a position controlled loop.
      .withFeedforward(new SimpleMotorFeedforward(0, 0, 0)) // Helps track changing RPM goals
      .withMotorInverted(false)
      .withTelemetry("LowerFlyWheel", SmartMotorControllerConfig.TelemetryVerbosity.HIGH);


  private SmartMotorController lowerFlyWheel = new TalonFXWrapper(new TalonFX(4),
                                                                  DCMotor.getKrakenX60(1),
                                                                  lowerFlyWheelConfig);

  private SmartMotorControllerConfig upperFlyWheelConfig = new SmartMotorControllerConfig(this)
      .withControlMode(ControlMode.CLOSED_LOOP)
      .withIdleMode(MotorMode.COAST)
//      .withWheelDiameter(Inches.of(4)) // Only needed to find the MPH of the flywheel for fun.
      .withGearing(new MechanismGearing(GearBox.fromReductionStages(3, 4)))
      .withMomentOfInertia(Inches.of(4), Pounds.of(2))
      .withClosedLoopController(1,
                                0,
                                0) // You generally do not want a profile because its not a position controlled loop.
      .withFeedforward(new SimpleMotorFeedforward(0, 0, 0)) // Helps track changing RPM goals
      .withMotorInverted(false)
      .withTelemetry("UpperFlyWheel", SmartMotorControllerConfig.TelemetryVerbosity.HIGH);
  private SmartMotorController       upperflyWheel       = new TalonFXWrapper(new TalonFX(6),
                                                                              DCMotor.getKrakenX60(1),
                                                                              upperFlyWheelConfig);


  public DoubleFlyWheelSubsystem()
  {
    upperflyWheel.setupTelemetry();
    lowerFlyWheel.setupTelemetry();
  }

  /**
   * Get the linear speed (commonly used for KM/h or MPH) of the flywheels.
   *
   * @return (Lower Flywheel Speed, Upper Flywheel Speed)
   */
  public Pair<LinearVelocity, LinearVelocity> getLinearSpeed()
  {
    return Pair.of(lowerFlyWheel.getMeasurementVelocity(), upperflyWheel.getMeasurementVelocity());
  }

  /**
   * Set the duty cycle of the upper and lower flywheels.
   *
   * @param lower Lower duty cycle.
   * @param upper Upper duty cycle.
   * @return {@link org.wpilib.command2.RunCommand}
   */
  public Command setDutyCycle(double lower, double upper)
  {
    return startRun(() -> {
      lowerFlyWheel.stopClosedLoopController();
      upperflyWheel.stopClosedLoopController();
    }, () -> {
      lowerFlyWheel.setDutyCycle(lower);
      upperflyWheel.setDutyCycle(upper);
    }).finallyDo(() -> {
      lowerFlyWheel.startClosedLoopController();
      upperflyWheel.startClosedLoopController();
    }).withName("Set Duty Cycle (Double FlyWheel)");
  }

  /**
   * Set the voltage of the upper and lower flywheels.
   *
   * @param lower Lower voltage.
   * @param upper Upper voltage.
   * @return {@link org.wpilib.command2.RunCommand}
   */
  public Command setVoltage(Voltage lower, Voltage upper)
  {
    return startRun(() -> {
      lowerFlyWheel.stopClosedLoopController();
      upperflyWheel.stopClosedLoopController();
    }, () -> {
      lowerFlyWheel.setVoltage(lower);
      upperflyWheel.setVoltage(upper);
    }).finallyDo(() -> {
      lowerFlyWheel.startClosedLoopController();
      upperflyWheel.startClosedLoopController();
    }).withName("Set Voltage (Double FlyWheel)");
  }

  /**
   * Create a {@link org.wpilib.command2.RunCommand} to set the speeds for the upper and lower flywheels
   * based off the distance to the goal.
   *
   * @param distanceToGoal Distance from the center of the robot to the goal on the XY plane.
   * @return {@link org.wpilib.command2.RunCommand}
   */
  public Command setSpeedForDistance(Supplier<Distance> distanceToGoal)
  {
    return run(() -> {
      var lowerSpeeds = RPM.of(DoubleFlyWheelConstants.distanceToRPM.getFirst()
                                                                    .get(distanceToGoal.get().in(Meters)));
      var upperSpeeds = RPM.of(DoubleFlyWheelConstants.distanceToRPM.getSecond()
                                                                    .get(distanceToGoal.get().in(Meters)));
      lowerFlyWheel.setVelocity(lowerSpeeds);
      upperflyWheel.setVelocity(upperSpeeds);
    }).withName("Set Speed For Distance (Double FlyWheel)");
  }

  /**
   * Set the velocity directly.
   *
   * @param lower Lower supplier
   * @param upper Upper velocity supplier
   * @return {@link org.wpilib.command2.RunCommand}
   */
  public Command setVelocity(Supplier<AngularVelocity> lower, Supplier<AngularVelocity> upper)
  {
    return run(() -> {
      lowerFlyWheel.setVelocity(lower.get());
      upperflyWheel.setVelocity(upper.get());
    });
  }

  /**
   * Set the velocity directly.
   *
   * @param lower Lower Velocity
   * @param upper Upper velocity
   * @return {@link org.wpilib.command2.RunCommand}
   */
  public Command setVelocity(AngularVelocity lower, AngularVelocity upper)
  {
    return run(() -> {
      lowerFlyWheel.setVelocity(lower);
      upperflyWheel.setVelocity(upper);
    });
  }

  @Override
  public void periodic()
  {
    lowerFlyWheel.updateTelemetry();
    upperflyWheel.updateTelemetry();
  }

  @Override
  public void simulationPeriodic()
  {
    lowerFlyWheel.simIterate();
    upperflyWheel.simIterate();
  }
}
